apps/els_lsp/src/els_completion_provider.erl

```diff
@@ -418,6 +427,77 @@ complete_type_definition(Document, Name, ItemFormat) ->
         modules(NameBinary) ++
         atoms(Document, NameBinary).
 
+-spec complete_function_clause(binary(), options()) -> items().
+complete_function_clause(Prefix, Opts) ->
+    #{
+        document := Document,
+        line := Line,
+        column := Column
+    } = Opts,
+
+    #{
+        pois := POIs,
+        text := Text
+    } = Document,
+    MatchedPois = match_all_pos(POIs, {Line, Column}),
+    MatchedFunctionPoi = lists:search(
+        fun(Poi) ->
+            case Poi of
+                #{kind := function} -> true;
+                _ -> false
+            end
+        end,
+        MatchedPois
+    ),
+
+    CompletionLine = els_text:line(Text, Line - 1),
+    %% Prefix does not include trigger character, but CompletionLine does
+    AtEndOfLine = size(Prefix) == (size(CompletionLine) - 1),
+
+    case {MatchedFunctionPoi, AtEndOfLine} of
+        %% We want to complete only at end of function clause body, to prevent
+        %% completing in middle of a clause in a case, receive, etc.
+        {{value, MatchedFun}, true} ->
+            #{data := #{clause_body_ranges := ClauseBodies}} = MatchedFun,
+
+            MatchedClauseBody =
+                lists:search(
+                    fun(#{to := {BodyEndLine, _Col}}) -> BodyEndLine == Line end, ClauseBodies
+                ),
+            case MatchedClauseBody of
+                {value, _} ->
+                    #{id := {Name, Arity}} = MatchedFun,
+                    ArgumentList = lists:map(
+                        fun(Num) ->
+                            io_lib:format("${~p:Arg~p}", [Num, Num])
+                        end,
+                        lists:seq(1, Arity)
+                    ),
+                    ArgumentText = lists:join(", ", ArgumentList),
+
+                    Snippet =
+                        io_lib:format(
+                            "\n~s(~s) ->\n    ${~p:Body}",
+                            [atom_to_list(Name), ArgumentText, Arity + 1]
+                        ),
+
+                    [
+                        #{
+                            label => <<"Add function clause">>,
+                            kind => ?COMPLETION_ITEM_KIND_SNIPPET,
+                            insertText => list_to_binary(Snippet),
+                            insertTextFormat => ?INSERT_TEXT_FORMAT_SNIPPET,
+                            insertTextMode => ?INSERT_TEXT_MODE_AS_IS
```

Comment:
Not 100% happy about the "As Is" insert mode, since I think it might not be perfect for people with other indentation styles..

Could it possibly be done using "textEdit" property of the completion item? But can that one add text on the line below?